# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1705605049 -3600
# Parent  f8a6d16403676f66f709b936451862773e8e5833
Feature: RegExp in search terms. Reported as https://bugzilla.mozilla.org/show_bug.cgi?id=19442 in 1999.
* * *
Feature: RegExp in search terms: Switch from std::regex to ICU's regex.

diff --git a/mailnews/search/public/nsMsgSearchTerm.h b/mailnews/search/public/nsMsgSearchTerm.h
--- a/mailnews/search/public/nsMsgSearchTerm.h
+++ b/mailnews/search/public/nsMsgSearchTerm.h
@@ -1,15 +1,32 @@
 /* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #ifndef __nsMsgSearchTerm_h
 #define __nsMsgSearchTerm_h
+
+// Using ICU's regex.
+// For this to link, the following files need to be moved from the
+// "unused" list to the "used" list in
+// mozilla-central/config/external/icu/i18n/sources.mozbuild
+// '/intl/icu/source/i18n/regexcmp.cpp'
+// '/intl/icu/source/i18n/regeximp.cpp'
+// '/intl/icu/source/i18n/regexst.cpp'
+// '/intl/icu/source/i18n/regextxt.cpp'
+// '/intl/icu/source/i18n/rematch.cpp'
+// '/intl/icu/source/i18n/repattrn.cpp'
+// Furthermore, this build option needs to be cleared:
+// # DEFINES['UCONFIG_NO_REGULAR_EXPRESSIONS'] = True
+#include "../../../../intl/icu/source/i18n/unicode/regex.h"
+
+using namespace U_ICU_NAMESPACE;  // currently icu_73
+
 //---------------------------------------------------------------------------
 // nsMsgSearchTerm specifies one criterion, e.g. name contains phil
 //---------------------------------------------------------------------------
 #include "nsIMsgSearchSession.h"
 #include "nsIMsgSearchScopeTerm.h"
 #include "nsIMsgSearchTerm.h"
 #include "nsIMsgSearchCustomTerm.h"
 
@@ -78,11 +95,16 @@ class nsMsgSearchTerm : public nsIMsgSea
    *
    * @param aValue  the string to switch
    */
   void ToLowerCaseExceptSpecials(nsACString& aValue);
   nsresult InitializeAddressBook();
   nsresult MatchInAddressBook(const nsAString& aAddress, bool* pResult);
   // fields used by search in address book
   nsCOMPtr<nsIAbDirectory> mDirectory;
+
+  // Used for Regex search.
+  nsAutoString mLastNeedle;
+  RegexMatcher* mRegex;
+  bool mRegexValid;
 };
 
 #endif
diff --git a/mailnews/search/src/nsMsgImapSearch.cpp b/mailnews/search/src/nsMsgImapSearch.cpp
--- a/mailnews/search/src/nsMsgImapSearch.cpp
+++ b/mailnews/search/src/nsMsgImapSearch.cpp
@@ -147,16 +147,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
@@ -165,16 +169,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntContain, 1);
@@ -182,16 +190,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntContain, 1);
@@ -199,16 +211,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntContain, 1);
@@ -216,16 +232,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntContain, 1);
@@ -233,25 +253,33 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntContain, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntContain, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Isnt, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Isnt, 1);
@@ -320,16 +348,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Keywords, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Keywords, nsMsgSearchOp::Contains, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Keywords, nsMsgSearchOp::DoesntContain, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Keywords, nsMsgSearchOp::DoesntContain, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Keywords, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetEnabled  (nsMsgSearchAttrib::Keywords, nsMsgSearchOp::Is, 1);
   m_offlineMailTable->SetAvailable(nsMsgSearchAttrib::Keywords, nsMsgSearchOp::Isnt, 1);
@@ -443,16 +475,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
@@ -461,16 +497,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntContain, 1);
@@ -478,16 +518,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntContain, 1);
@@ -495,16 +539,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntContain, 1);
@@ -512,16 +560,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntContain, 1);
@@ -529,16 +581,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
 
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Isnt, 1);
@@ -590,16 +646,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
   m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_onlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
 
   return rv;
 }
 
 nsresult
 nsMsgSearchValidityManager::InitOfflineMailFilterTable()
 {
   NS_ASSERTION(!m_offlineMailFilterTable, "offline mail filter table already initted");
@@ -613,16 +673,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
@@ -631,16 +695,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntContain, 1);
@@ -648,16 +716,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntContain, 1);
@@ -665,16 +737,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntContain, 1);
@@ -682,16 +758,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntContain, 1);
@@ -699,25 +779,33 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Contains, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntContain, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntContain, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Isnt, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Body, nsMsgSearchOp::DoesntMatch, 1);
 
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Isnt, 1);
@@ -774,16 +862,20 @@ nsMsgSearchValidityManager::InitOfflineM
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
   m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_offlineMailFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
+  m_offlineMailFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
 
   return rv;
 }
 
 // Online Manual is used for IMAP and NEWS, where at manual
 // filtering we have junk info, but cannot assure that the
 // body is available.
 nsresult
@@ -800,16 +892,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Contains, 1);
@@ -818,16 +914,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::To, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntContain, 1);
@@ -835,16 +935,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::CC, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntContain, 1);
@@ -852,16 +956,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::ToOrCC, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntContain, 1);
@@ -869,16 +977,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::DoesntMatch, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsInAB, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::AllAddresses, nsMsgSearchOp::IsntInAB, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntContain, 1);
@@ -886,16 +998,20 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
 
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Isnt, 1);
@@ -980,12 +1096,16 @@ nsMsgSearchValidityManager::InitOnlineMa
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
   m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  m_onlineManualFilterTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
+  m_onlineManualFilterTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
 
   return rv;
 }
 // clang-format on
diff --git a/mailnews/search/src/nsMsgLocalSearch.cpp b/mailnews/search/src/nsMsgLocalSearch.cpp
--- a/mailnews/search/src/nsMsgLocalSearch.cpp
+++ b/mailnews/search/src/nsMsgLocalSearch.cpp
@@ -805,16 +805,20 @@ nsresult SetLocalNews(nsIMsgSearchValidi
   aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Isnt, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
+  aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+  aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
+  aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsInAB, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::IsntInAB, 1);
 
   aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntContain, 1);
@@ -822,16 +826,20 @@ nsresult SetLocalNews(nsIMsgSearchValidi
   aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Isnt, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
+  aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+  aTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
+  aTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
 
   aTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsBefore, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::IsAfter, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::Date, nsMsgSearchOp::Is, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::Date, nsMsgSearchOp::Isnt, 1);
@@ -869,16 +877,20 @@ nsresult SetLocalNews(nsIMsgSearchValidi
   aTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Isnt, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
   aTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
   aTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
+  aTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  aTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+  aTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
+  aTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
   return NS_OK;
 }
 // clang-format on
 
 nsresult nsMsgSearchValidityManager::InitLocalNewsTable() {
   NS_ASSERTION(nullptr == m_localNewsTable,
                "already have local news validity table");
   nsresult rv = NewTable(getter_AddRefs(m_localNewsTable));
diff --git a/mailnews/search/src/nsMsgSearchNews.cpp b/mailnews/search/src/nsMsgSearchNews.cpp
--- a/mailnews/search/src/nsMsgSearchNews.cpp
+++ b/mailnews/search/src/nsMsgSearchNews.cpp
@@ -345,41 +345,53 @@ nsresult nsMsgSearchValidityManager::Ini
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Contains, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Contains, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Is, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::BeginsWith, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::EndsWith, 1);
+    m_newsTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+    m_newsTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::Matches, 1);
+    m_newsTable->SetAvailable(nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
+    m_newsTable->SetEnabled  (nsMsgSearchAttrib::Sender, nsMsgSearchOp::DoesntMatch, 1);
 
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Contains, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Is, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::BeginsWith, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::EndsWith, 1);
+    m_newsTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+    m_newsTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::Matches, 1);
+    m_newsTable->SetAvailable(nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
+    m_newsTable->SetEnabled  (nsMsgSearchAttrib::Subject, nsMsgSearchOp::DoesntMatch, 1);
 
 #if 0
     // Size should be handled after the fact...
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Size, nsMsgSearchOp::IsGreaterThan, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Size, nsMsgSearchOp::IsGreaterThan, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::Size, nsMsgSearchOp::IsLessThan, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::Size, nsMsgSearchOp::IsLessThan, 1);
 #endif
     m_newsTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Contains, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Contains, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Is, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::BeginsWith, 1);
     m_newsTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
     m_newsTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::EndsWith, 1);
+    m_newsTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+    m_newsTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::Matches, 1);
+    m_newsTable->SetAvailable(nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
+    m_newsTable->SetEnabled  (nsMsgSearchAttrib::OtherHeader, nsMsgSearchOp::DoesntMatch, 1);
     // clang-format on
   }
 
   return rv;
 }
 
 nsresult nsMsgSearchValidityManager::InitNewsFilterTable() {
   NS_ASSERTION(nullptr == m_newsFilterTable,
diff --git a/mailnews/search/src/nsMsgSearchTerm.cpp b/mailnews/search/src/nsMsgSearchTerm.cpp
--- a/mailnews/search/src/nsMsgSearchTerm.cpp
+++ b/mailnews/search/src/nsMsgSearchTerm.cpp
@@ -325,16 +325,17 @@ nsMsgSearchTerm::nsMsgSearchTerm() {
   // initialize this to zero
   m_value.attribute = 0;
   m_value.u.priority = 0;
   m_attribute = nsMsgSearchAttrib::Default;
   m_operator = nsMsgSearchOp::Contains;
   mBeginsGrouping = 0;
   mEndsGrouping = 0;
   m_matchAll = false;
+  mRegex = nullptr;
 
   // valgrind warning during GC/java data check suggests
   // m_booleanp needs to be initialized too.
   m_booleanOp = nsMsgSearchBooleanOp::BooleanAND;
 }
 
 nsMsgSearchTerm::nsMsgSearchTerm(nsMsgSearchAttribValue attrib,
                                  nsMsgSearchOpValue op, nsIMsgSearchValue* val,
@@ -350,19 +351,22 @@ nsMsgSearchTerm::nsMsgSearchTerm(nsMsgSe
   } else if (attrib == nsMsgSearchAttrib::Custom) {
     m_customId = aCustomString;
   }
 
   nsMsgResultElement::AssignValues(val, &m_value);
   m_matchAll = false;
   mBeginsGrouping = false;
   mEndsGrouping = false;
+  mRegex = nullptr;
 }
 
-nsMsgSearchTerm::~nsMsgSearchTerm() {}
+nsMsgSearchTerm::~nsMsgSearchTerm() {
+  if (mRegex) delete mRegex;
+}
 
 NS_IMPL_ISUPPORTS(nsMsgSearchTerm, nsIMsgSearchTerm)
 
 // Perhaps we could find a better place for this?
 // Caller needs to free.
 /* static */ char* nsMsgSearchTerm::EscapeQuotesInStr(const char* str) {
   int numQuotes = 0;
   for (const char* strPtr = str; *strPtr; strPtr++)
@@ -637,17 +641,18 @@ nsresult nsMsgSearchTerm::MatchArbitrary
     bool* pResult) {
   NS_ENSURE_ARG_POINTER(pResult);
 
   *pResult = false;
   nsresult rv = NS_OK;
   bool matchExpected = m_operator == nsMsgSearchOp::Contains ||
                        m_operator == nsMsgSearchOp::Is ||
                        m_operator == nsMsgSearchOp::BeginsWith ||
-                       m_operator == nsMsgSearchOp::EndsWith;
+                       m_operator == nsMsgSearchOp::EndsWith ||
+                       m_operator == nsMsgSearchOp::Matches;
   // Initialize result to what we want if we don't find the header at all.
   bool result = !matchExpected;
 
   nsCString dbHdrValue;
   msg->GetStringProperty(m_arbitraryHeader.get(), dbHdrValue);
   if (!dbHdrValue.IsEmpty()) {
     // Match value with the other info. It doesn't check all header occurrences,
     // so we use it only if we match and do line by line headers parsing
@@ -1019,16 +1024,49 @@ nsresult nsMsgSearchTerm::MatchString(co
                            nsCaseInsensitiveStringComparator))
         result = true;
       break;
     case nsMsgSearchOp::EndsWith:
       if (StringEndsWith(utf16StrToMatch, needle,
                          nsCaseInsensitiveStringComparator))
         result = true;
       break;
+    case nsMsgSearchOp::Matches:
+    case nsMsgSearchOp::DoesntMatch: {
+      if (!needle.Equals(mLastNeedle) || !mRegex) {
+        mRegexValid = true;
+        if (mRegex) delete mRegex;
+        UErrorCode status = U_ZERO_ERROR;
+        mRegex = new RegexMatcher(static_cast<const char16_t*>(needle.get()),
+                                  UREGEX_CASE_INSENSITIVE, status);
+        if (U_FAILURE(status)) {
+          // Even on failure a new `RegexMatcher` object is returned,
+          // so no special case required.
+          NS_WARNING("RegexMatcher. Likely a bad regular expression.");
+          mRegexValid = false;
+        }
+        mLastNeedle = needle;
+      }
+      if (!mRegexValid) {
+        *pResult = false;
+        return NS_OK;
+      }
+      // We need to use a local variable here since `reset()` doesn't take a
+      // copy of the string, so it needs to survive the invocation.
+      const UnicodeString string(static_cast<const char16_t*>(
+          PromiseFlatString(utf16StrToMatch).get()));
+      mRegex->reset(string);
+
+      bool found = mRegex->find();
+      if ((m_operator == nsMsgSearchOp::Matches && found) ||
+          (m_operator == nsMsgSearchOp::DoesntMatch && !found)) {
+        result = true;
+      }
+      break;
+    }
     default:
       rv = NS_ERROR_FAILURE;
       NS_ERROR("invalid compare op for matching search results");
   }
 
   *pResult = result;
   return rv;
 }
